python-sdk/nuscenes/predict/input_representation/agents.py

```diff
@@ -0,0 +1,279 @@
+# nuScenes dev-kit.
+# Code written by Freddy Boulton, 2020.
+
+from typing import Any, Dict, List, Tuple, Callable
+import colorsys
+
+import cv2
+import numpy as np
+from pyquaternion import Quaternion
+
+from nuscenes.predict import PredictHelper
+from nuscenes.predict.helper import (angle_of_rotation, quaternion_yaw)
+from nuscenes.predict.input_representation.interface import AgentRepresentation
+from nuscenes.predict.input_representation.utils import convert_to_pixel_coords, get_crops
+
+History = Dict[str, List[Dict[str, Any]]]
+
+def pixels_to_box_corners(row_pixel: int, column_pixel: int,
+                          length_in_pixels: float,
+                          width_in_pixels: float,
+                          yaw_in_radians: float) -> np.ndarray:
+    """
+    Computes four corners of 2d bounding box for agent.
+    The coordinates of the box are in pixels.
+    :param row_pixel: Row pixel of the agent.
+    :param column_pixel: Column pixel of the agent.
+    :param length_in_pixels: Length of the agent.
+    :param width_in_pixels: Width of the agent.
+    :param yaw_in_radians: Yaw of the agent (global coordinates).
+    :return: numpy array representing the four corners of the agent.
+    """
+
+    # cv2 has the convention where they flip rows and columns so it matches
+    # the convention of x and y on a coordinate plane
+    # Also, a positive angle is a clockwise rotation as opposed to counterclockwise
+    # so that is why we negate the rotation angle
+    coord_tuple = ((column_pixel, row_pixel), (length_in_pixels, width_in_pixels), -yaw_in_radians * 180 / np.pi)
+
+    box = cv2.boxPoints(coord_tuple)
+
+    return box
+
+
+def get_track_box(annotation: Dict[str, Any],
+                  center_coordinates: Tuple[int, int],
+                  center_pixels: Tuple[int, int],
+                  resolution: float = 0.1) -> np.ndarray:
+    """
+    Get four corners of bounding box for agent in pixels.
+    :param annotation: The annotation record of the agent.
+    :param ego_coordinates: (x, y) coordinates in global frame
+        of the center of the image
+    :param center_pixels: (row_index, column_index) location of the center
+        of the image in pixel coordinates.
+    :param resolution: Resolution pixels/meter of the image.
+    """
+
+    location = annotation['translation'][:2]
+    yaw_in_radians = quaternion_yaw(Quaternion(annotation['rotation']))
+
+    row_pixel, column_pixel = convert_to_pixel_coords(location,
+                                                      center_coordinates,
+                                                      center_pixels, resolution)
+
+    width = annotation['size'][0] / resolution
```

Comment:
Thoughts on adding a check for zero resolution?

Reply:
Done